docs/proposals/devfile/outer-loop-build-and-deploy.md

```diff
@@ -0,0 +1,310 @@
+# Project Proposal - Devfile 2.0 outer-loop devfile build and deploy functions
+
+## Background
+Existing devfile 2.0 mainly focuses on providing inner-loop support. This proposal adds support on outer-loop scenarios to expand the scope of usage of devfile support.
+
+### Outer-loop scenario
+As an application developer, I would like to build a microservice and deploy it to Kubernetes to do a build that is typically done as part of a pipeline
+
+- Do a full build to build the container image
+    - Input\
+    The build may use different technologies for building, e.g. dockerfile, buildpacks
+    - Output:\
+    A container image that contains the microservice and ready to be deployed to Kubernetes
+- Deploy the built image to Kubernetes:
+    - Deploy the built image to a cluster:\
+    Deployment may use different methods of deployment, e.g. Kubernetes deployment manifest, Operators, Helm, Knative
+
+The focus of this proposal is to provide the ability to build and deploy applications or services in a way similar to what will be done for production. In contrast to the inner-loop scenario, the build will not only focus on the building of the application but will also cover building the runtime container that runs the application.
+
+We will cover two main stages:
+1. Build the image
+2. Deployment of the image. Deploy the image to a cluster
+
+## Build the image
+
+This includes the creation of the base image of the runtime container, building the application, and packaging the application as part of the container image. Different mechanisms can be used for building the image, for example:
+1. Dockerfile
+2. Dockerfile stages
+3. Buildpacks
+4. Source-to-Image (S2I)
+5. Tekton
+
+The initial design will focus on the first two items for discussion. Other ways of building the image will be covered later.
+
+The definition of the image build introduces a new component type `image` for the definition of the image build. The `image` component definition can use different build strategies, e.g. `dockerfile`, `s2i`, `buildpacks`, etc.
+
+__Alternative approach:__ we can consider reusing the `exec` type except that these build types are special and we may not want to incorporate the full image build command into an `exec` command. Therefore, I chose to use a different type of command here.
+
+### Example using a file Dockerfile as part of the app:
+    variables:
+      myimage: myimagename
+    components:
+      - name: mydockerfileimage
+        image:
+          imageName: {{myimage}}
+          dockerfile:
+            buildContext: ${PROJECTS_ROOT}/build
+            location: Dockerfile
+            args: [ "arg1", "arg2", "arg3" ]
+            rootRequired: false
```

Comment:
Use regular code blocks to get syntax highlighting.
```suggestion
```yaml
    variables:
      myimage: myimagename
    components:
      - name: mydockerfileimage
        image:
          imageName: {{myimage}}
          dockerfile:
            buildContext: ${PROJECTS_ROOT}/build
            location: Dockerfile
            args: [ "arg1", "arg2", "arg3" ]
            rootRequired: false
```
```

Reply:
fixed